lambdaclass/mirra_backend
# [GH-531] Add google JWT validation

## Motivation
We need a way to handle the google sign in validation from the backend server.
Closes #531 

## Changes
- [Add secret_key_base to gateway endpoint](https://github.com/lambdaclass/mirra_backend/pull/532/commits/3d3841c804091d2b550e2da8383e300d43e7de17)
- [Add GoogleUser table and schema](https://github.com/lambdaclass/mirra_backend/pull/532/commits/c213623347b882aa3b669f30f8c8ef6dfb3e3e41)
- [Add find and create functions for GoogleUser](https://github.com/lambdaclass/mirra_backend/pull/532/commits/220b881a996b7e8550a5041ce26ce25958516e48)
- [Add find_or_create_google_user_by_email/1 function doc](https://github.com/lambdaclass/mirra_backend/pull/532/commits/f2dffb238061a38caff24d12e5046d3270fab806)
- [Add Joken and GoogleCertificates dependencies](https://github.com/lambdaclass/mirra_backend/pull/532/commits/98a7f873287b54e8c5db2833c9d91e509c5eecb5)
- [Add Joken hook to fetch google public keys and create a joken signer …](https://github.com/lambdaclass/mirra_backend/pull/532/commits/7b9761d743e6a695ef40d8a702ce1ba3064662d3)
- [Add GoogleTokenManager module to handle tokens verification and valid…](https://github.com/lambdaclass/mirra_backend/pull/532/commits/47a5e9b7fb89470ff6279052fe8a745a545003a6)
- [Start google certificates fetcher genserver with the application](https://github.com/lambdaclass/mirra_backend/pull/532/commits/0f99bfeae1ded49cbf921b3dfcffbaa7c8d463d2)
- [Add validate token endpoint](https://github.com/lambdaclass/mirra_backend/pull/532/commits/28b5aa319940d2f1304ff30442a4cc0d76c2a1b7)

## Docs
- [Google documentation](https://developers.google.com/identity/sign-in/web/backend-auth)
- Dependencies added:
  - [Joken](https://hexdocs.pm/joken/introduction.html) -> Dependency to verify and validate JWT.
  - [Google Certificates](https://hexdocs.pm/google_certs/readme.html) -> Dependency to fetch google JWK (public keys to validate signature).

## How to test it?
- Just get a tokenID and do a GET request to `http://localhost:4001/auth/google/token/:token_id`
- To get a tokenID ->
  - Create a google cloud dummy credential at https://console.cloud.google.com/.
  - Open cloud shell (see attached image).
  - Use the following command: `gcloud auth print-identity-token`
<img width="245" alt="Screenshot 2024-04-30 at 5 01 15 PM" src="https://github.com/lambdaclass/mirra_backend/assets/65268603/9f519dce-0205-4e62-a760-7191aca9b475">
